File: doubly_linked_list/doubly_linked_list.py
class Node:
    __slots__ = ("value", "prev", "next")

    def __init__(self, value=None):
        self.value = value
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        # Sentinel nodes (Always present, never removed)
        self.dummy_head = Node()
        self.dummy_tail = Node()
        self.dummy_head.next = self.dummy_tail
        self.dummy_tail.prev = self.dummy_head

    def insert_front(self, value):
        new = Node(value)
        old_first = self.dummy_head.next
        new.prev = self.dummy_head
        new.next = old_first
        self.dummy_head.next = new
        old_first.prev = new
        return new

    def insert_back(self, value):
        new = Node(value)
        old_last = self.dummy_tail.prev
        new.prev = old_last
        new.next = self.dummy_tail
        old_last.next = new
        self.dummy_tail.prev = new
        return new

    def delete(self, node):
        prev_node = node.prev
        next_node = node.next
        prev_node.next = next_node
        next_node.prev = prev_node
        node.prev = None
        node.next = None
        return node.value

    def traverse_forward(self):
        values = []
        current = self.dummy_head.next
        while current is not self.dummy_tail:
            values.append(current.value)
            current = current.next
        return values

    def traverse_backward(self):
        values = []
        current = self.dummy_tail.prev
        while current is not self.dummy_head:
            values.append(current.value)
            current = current.prev
        return values

    def reverse(self):
        current = self.dummy_head
        while current is not None:
            current.prev, current.next = current.next, current.prev
            current = current.prev
        self.dummy_head, self.dummy_tail = self.dummy_tail, self.dummy_head

    def __repr__(self):
        return f"DoublyLinkedList({self.traverse_forward()})"


def main():
    dll = DoublyLinkedList()
    dll.insert_front(1)
    dll.insert_front(2)
    dll.insert_back(3)
    print(dll.traverse_forward())  # expect [2, 1, 3]
    print(dll.traverse_backward())  # expect [3, 1, 2]

    dll.reverse()
    print(dll.traverse_forward())  # expect [3, 1, 2]
    print(dll.traverse_backward())  # expect [2, 1, 3]

    node = dll.insert_front(99)
    print(dll.traverse_forward())  # expect [99, 3, 1, 2]
    dll.delete(node)
    print(dll.traverse_forward())  # expect [3, 1, 2]


if __name__ == "__main__":
    main()

File: doubly_linked_list/doubly_linked_list.ml
(* Doubly Linked List — OCaml implementation
   Complexity: insert/delete O(1), iter/fold O(n), length O(n)
   Tricky part: mutable option-based prev/next pointers require careful
   pattern matching; no sentinel nodes — we use option types idiomatically.
   Improvement: could track length with a mutable counter for O(1) length. *)

type 'a node = {
  value : 'a;
  mutable prev : 'a node option;
  mutable next : 'a node option;
}

type 'a t = {
  mutable head : 'a node option;
  mutable tail : 'a node option;
}

let create () : 'a t = { head = None; tail = None }

let is_empty (dll : 'a t) : bool = dll.head = None

let push_front (dll : 'a t) (value : 'a) : 'a node =
  let node = { value; prev = None; next = dll.head } in
  (match dll.head with
   | Some old_head -> old_head.prev <- Some node
   | None -> dll.tail <- Some node);
  dll.head <- Some node;
  node

let push_back (dll : 'a t) (value : 'a) : 'a node =
  let node = { value; prev = dll.tail; next = None } in
  (match dll.tail with
   | Some old_tail -> old_tail.next <- Some node
   | None -> dll.head <- Some node);
  dll.tail <- Some node;
  node

let remove (dll : 'a t) (node : 'a node) : 'a =
  (match node.prev with
   | Some p -> p.next <- node.next
   | None -> dll.head <- node.next);
  (match node.next with
   | Some n -> n.prev <- node.prev
   | None -> dll.tail <- node.prev);
  node.prev <- None;
  node.next <- None;
  node.value

let peek_front (dll : 'a t) : 'a option =
  Option.map (fun n -> n.value) dll.head

let peek_back (dll : 'a t) : 'a option =
  Option.map (fun n -> n.value) dll.tail

let pop_front (dll : 'a t) : 'a option =
  Option.map (fun node -> remove dll node) dll.head

let pop_back (dll : 'a t) : 'a option =
  Option.map (fun node -> remove dll node) dll.tail

(* Fold left-to-right (head to tail) *)
let fold_left (f : 'acc -> 'a -> 'acc) (init : 'acc) (dll : 'a t) : 'acc =
  let rec go acc = function
    | None -> acc
    | Some node -> go (f acc node.value) node.next
  in
  go init dll.head

(* Fold right-to-left (tail to head) *)
let fold_right (f : 'a -> 'acc -> 'acc) (dll : 'a t) (init : 'acc) : 'acc =
  let rec go acc = function
    | None -> acc
    | Some node -> go (f node.value acc) node.prev
  in
  go init dll.tail

let iter (f : 'a -> unit) (dll : 'a t) : unit =
  fold_left (fun () x -> f x) () dll

let to_list (dll : 'a t) : 'a list =
  fold_right (fun x acc -> x :: acc) dll []

let to_list_rev (dll : 'a t) : 'a list =
  fold_left (fun acc x -> x :: acc) [] dll

let length (dll : 'a t) : int =
  fold_left (fun n _ -> n + 1) 0 dll

let reverse (dll : 'a t) : unit =
  let rec go = function
    | None -> ()
    | Some node ->
      let tmp = node.next in
      node.next <- node.prev;
      node.prev <- tmp;
      go tmp
  in
  go dll.head;
  let tmp = dll.head in
  dll.head <- dll.tail;
  dll.tail <- tmp

(* Move an existing node to the front — useful for LRU caches *)
let move_to_front (dll : 'a t) (node : 'a node) : unit =
  ignore (remove dll node);
  node.next <- dll.head;
  node.prev <- None;
  (match dll.head with
   | Some old_head -> old_head.prev <- Some node
   | None -> dll.tail <- Some node);
  dll.head <- Some node

let pp (pp_elt : Format.formatter -> 'a -> unit) (fmt : Format.formatter) (dll : 'a t) : unit =
  Format.fprintf fmt "[";
  let first = ref true in
  iter (fun x ->
    if !first then first := false
    else Format.fprintf fmt "; ";
    pp_elt fmt x
  ) dll;
  Format.fprintf fmt "]"

(* ── Demo ── *)
let () =
  let dll = create () in
  let _ = push_front dll 1 in
  let _ = push_front dll 2 in
  let _ = push_back dll 3 in

  Printf.printf "Forward:  ";
  Format.printf "%a\n" (pp Format.pp_print_int) dll;
  (* expect [2; 1; 3] *)

  Printf.printf "Backward: [%s]\n"
    (String.concat "; " (List.map string_of_int (to_list_rev dll)));
  (* expect [3; 1; 2] *)

  reverse dll;
  Printf.printf "Reversed: ";
  Format.printf "%a\n" (pp Format.pp_print_int) dll;
  (* expect [3; 1; 2] *)

  let n99 = push_front dll 99 in
  let n100 = push_back dll 100 in
  Printf.printf "Added:    ";
  Format.printf "%a\n" (pp Format.pp_print_int) dll;
  (* expect [99; 3; 1; 2; 100] *)

  ignore (remove dll n99);
  ignore (remove dll n100);
  Printf.printf "Removed:  ";
  Format.printf "%a\n" (pp Format.pp_print_int) dll;
  (* expect [3; 1; 2] *)

  (* move_to_front demo *)
  let node_middle = Option.get (Option.bind dll.head (fun n -> n.next)) in
  move_to_front dll node_middle;
  Printf.printf "Move mid: ";
  Format.printf "%a\n" (pp Format.pp_print_int) dll;
  (* expect [1; 3; 2] *)

  (* pop demo *)
  Printf.printf "Pop front: %d\n" (Option.get (pop_front dll));
  Printf.printf "Pop back:  %d\n" (Option.get (pop_back dll));
  Printf.printf "Remaining: ";
  Format.printf "%a\n" (pp Format.pp_print_int) dll;
  Printf.printf "Length: %d\n" (length dll)
